### ISSUE TITLE
Remote console/chat via WebSocket

### ISSUE THREAD

**Issue by Mqxx** [itzg/docker-minecraft-server#3152]
### Enhancement Type

A completely new feature

### Describe the enhancement

Currently, the only way to connect to the server and send commands is to use the rcon-cli that your container provides. Rcon kinda sucks... And I mean specifically Rcon, **not your implementation of rcon-cli**. There is currently no better way to do it.

The problem with Rcon is that you can only retrieve the console output of the command you just ran, not the entire console output. This makes it difficult, for example, to build a WebUI or other applications that require the entire output. I think this is also the reason why many people currently rely on technologies like [Screen](https://www.gnu.org/software/screen/manual/screen.html) or [Tmux](https://github.com/tmux/tmux) (which I also used before). Simply to get the live console.

It would be really great if you could implement a WebSocket hook that directly accesses the console in the server from the container so that you can get a live stream of the console and also send commands via WebSocket.

This way it would be possible to build better management tools that use this container to manage the server.

Thanks for consideration 🙂

**Comment by itzg** [comment]
I agree with the frustrations with RCON, but then it's just doing the command-response that it's designed for. 

A websocket interface makes sense especially since docker and kubernetes use that to enable programmatic access to interact with containers. Speaking of which, keep in mind that the output of the Minecraft server can always be remotely retrieved via docker/kubernetes logs. 

It could be done similar to https://github.com/itzg/mc-server-runner/pull/56

To set expectations, unless someone contributes this feature, I can't guarantee I'll add this any time soon or ever.

**Comment by Mqxx** [comment]
I will look into this. My idea was to implement a WebSocket server using for example Deno (JS/TS runntime) and watching `latest.log` using the build-in file watcher.

Sending commands would be done by sending it through WebSocket to the WebSocketServer and then the server would send it via Rcon to the Minecraft server.

Receiving commands would be done by listening on the `latest.log` for changes and if there are any, then the changes are send via WebSocket to the client.

That's currently the ony way (that I know) on how to "send and recive as raw as possible".

An other way would be to go directly with a Fabric mod (at least for my purpose, cause I use Fabric) that hooks into the log of Minecraft.

**Comment by Mqxx** [comment]
Just for tracking on which features would be nice to have for the WebSocket connection:
- [x] Sending commands
- [x] Receiving commands
- [ ] Query available commands
- [ ] Query possible auto completions for provided string (Ex. Sending "hel" would result in a response list like ["help", ...], but this could be implemented client side)
- [ ] ~~(Optional: Query server stats, like CPU and RAM usage)~~ (Out of scope)

**Comment by Mqxx** [comment]
I have written a mod ([`msmp-console`](https://github.com/MinecraftPlayground/msmp-console-mod)) that extends the default MSMP (Minecraft Server Management Protocol) to allow notifications to be send for every console log event. This covers most of my requirements. I will extend this mod to also be able to send commands.

**Comment by itzg** [comment]
Just FYI, websocket access was added, as described here

https://docker-minecraft-server.readthedocs.io/en/latest/sending-commands/websocket/

**Comment by Mqxx** [comment]
> Just FYI, websocket access was added, as described here
> 
> https://docker-minecraft-server.readthedocs.io/en/latest/sending-commands/websocket/

Nice, yes I have seen this, but this looks for me like reinventing the wheel. Minecraft does already provide the MSMP for newer versions. The WebSocket solution that you implemented for newer versions is also great! In case you want to use older versions of Minecraft that do not support MSMP or non Fabric servers.

That's great and all, but I think I will stick to the MSMP for future development. It looks like the more promising implementation to me 🙂, also since it is naively implemented for all future Minecraft versions by default.
